### PR TITLE
Feature/eslint checker plugin

### PR DESCRIPTION
**Pull request by TimLehner**
Inspired by the existing Typescript Checker Plugin. This checker is a more bare-bones version of the typesript plugin, and doesn't have much in the way of customisability or testing yet. 

I have this working for my personal project which uses next-js+typescript, but I haven't tried this using other projects like create-react-app or bundlers like webpack.

I am also unsure what the best way to add the types for `StrykerOptions`, I wish to add a `lintConfigFile?: string`, but I only saw a big interface defined in src_generated which I thought I shouldn't touch.

Finally apologies I didn't open an issue before, I only just started on this earlier today and it all came together quite quickly

### CODE REVIEW

**Comment by TimLehner** [comment]
Progress:

- [x] Implementation
- [x] unit tests
- [x] integration tests
- [x] e2e tests:
  - [x] mocha-javascript-eslint
  - [ ] jest-js-eslint
  - [ ] jest-ts-eslint
  - [ ] jest-old-version-eslint
  - [ ] other e2e tests?

**Comment by nicojs** [comment]
I don't think it is needed to have an eslint e2e test for each test runner. My reasoning is: checkers are pretty isolated from test runners. The added value is minimal.

Rather I'm now thinking the following e2e tests:

- A JS project with eslint recommended rule set
- A JS project with [eslint type-aware recommended](https://typescript-eslint.io/docs/linting/type-linting/) rules active 
- A TS project with the typescript-checker AND [eslint type-aware recommended](https://typescript-eslint.io/docs/linting/type-linting/) rules active.

@TimLehner what do you think?

**Comment by TimLehner** [comment]
Yes, that sounds much better 👍

**Comment by TimLehner** [comment]
@nicojs I have implemented the requested e2e tests. I think this is now ready for a proper review

**Comment by nicojs** [comment]
@TimLehner I've got time to look at this tomorrow and next Monday. The build is failing because of some dependency stuff, which I've fixed yesterday on master. Would you mind merging those changes in your branch?

**Comment by nicojs** [comment]
Thanks for the push rights! Will be looking at this today 👀

File: packages/eslint-checker/README.md
Comment (simondel) [inline packages/eslint-checker/README.md:36]: I think it would also be smart to disable https://eslint.org/docs/rules/max-len due to the fact that mutation switching will result in (much) longer lines of code

File: packages/eslint-checker/package.json
Comment (simondel) [inline packages/eslint-checker/package.json:18]: This url (and other urls) will have to be changed prior to merging

File: packages/eslint-checker/package.json
Comment (simondel) [inline packages/eslint-checker/package.json:36]: I can't find typescript as a direct dependency in your source files. Am I overlooking something or is can this be part of the devDependencies?

File: packages/eslint-checker/package.json
Comment (simondel) [inline packages/eslint-checker/package.json:34]: @nicojs (how) do we want to set these dependency version? As any part of this major version? Or fixed to a minor/patch version?

File: packages/eslint-checker/src/esconfig-helpers.ts
Comment (simondel) [inline packages/eslint-checker/src/esconfig-helpers.ts:28]: I think this if check can be removed:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete#description

>If the property which you are trying to delete does not exist, delete will not have any effect and will return true.

File: packages/eslint-checker/src/result-helpers.ts
Comment (simondel) [inline packages/eslint-checker/src/result-helpers.ts:9]: I think the `_` can be removed in the param name `_result`

**Review by nicojs — CHANGES_REQUESTED** [review]
This looks clean and promising. Still some work to be done, though. I agree with @simondel 's remarks unless I say so in the comments

File: packages/eslint-checker/package.json
Comment (nicojs) [reply packages/eslint-checker/package.json:36]: I think `typescript` is added here as a transient dependency. You need it, if you use the `HybridFileSystem`. Not the best solution IMHO, see my remarks there.

File: packages/eslint-checker/schema/eslint-checker-options.json
Comment (TimLehner) [reply packages/eslint-checker/schema/eslint-checker-options.json:7]: I'm not entirely sure I understand what you mean, so I've not addressed this one yet

File: packages/util/src/fs/hybrid-file-system.ts
Comment (TimLehner) [reply packages/util/src/fs/hybrid-file-system.ts:3]: I think I lean towards having slightly duplicated code, and keeping this `fs` `async`? In which case I'll revert the extraction to utils and use `fs.readFile`

File: packages/eslint-checker/src/eslint-checker.ts
Comment (TimLehner) [reply packages/eslint-checker/src/eslint-checker.ts:45]: This sounds useful, I'll hold of on making any changes here before seeing that PR

File: packages/eslint-checker/README.md
Comment (TimLehner) [reply packages/eslint-checker/README.md:36]: Yes, I was thinking of getting a minimal set of rules in initially, and refining this later. Happy to add any rules you think are sensible before merging this PR

**Comment by TimLehner** [comment]
I think I've addressed the majority of your comments, let me know if I've missed anything

**Comment by nicojs** [comment]
I've was _finally_ able to merge # which allows us to progress here. I've updated the code, file descriptions are now being passed into the constructor.

Also did some more work:
- rename e2e tests to start with the name "eslint", that makes more sense IMO (i was looking for 'eslint')
- Ive added a e2e test for the validation schema
- I went ahead and changed the config setting to be grouped under `eslint`. That allows us to expand on eslint settings in the future. This is also what we did for mocha, karma, etc. The `tsconfigFile` is an exception, because Stryker's core also uses that setting.
- Ive added the vscode config to easily run the tests from vscode

I was looking at the report in the e2e tests, and was not happy with the result TBH:

eslint-false-negatives

As you can see, there are 3 linting errors here. However, these are all false negatives in my opinion. Maybe its better to _whitelist_ certain rules, instead of the blacklisting approach.

@TimLehner what do you think about that?

File: packages/util/src/fs/hybrid-file-system.ts
Comment (nicojs) [reply packages/util/src/fs/hybrid-file-system.ts:3]: I agree! Feel free to revert 🙏

**Comment by TimLehner** [comment]
Hmm, I think you make a good point on the false negative there, whitelisting would be useful, then if there is a specific surviving mutant you dismiss due to your lint rules, you can specifically enable the relevant rule and use Stryker as normal. I do feel the current default settings are too aggressive and kill too many mutants that are valid cases of lacking tests

**Comment by nicojs** [comment]
Yeah, what rules would you say are really useful to enable? What was the reason for you to embark on this journey?

**Comment by TimLehner** [comment]
Apologies for the delay in my response, I've been distracted.

It's tricky, since my initial motivation were a number of rather niche react-specific ESLint rules, namely: [react-hooks/exhaustive-deps](https://www.npmjs.com/package/eslint-plugin-react-hooks)

Although I have also found it too be potentially too aggressive there. I'm not sure what your thoughts are on this use-case?

**Pull request closed without merging**
